### PR TITLE
Support Azure OOT credential provider

### PR DESCRIPTION
**Pull request by lzhecheng**
For Linux and Windows, OOT credential provider and its config will be downloaded.

What this PR does / why we need it:
Support Azure OOT credential provider

Which issue(s) this PR fixes (optional, in fixes #<issue number>(, fixes #<issue_number>, ...) format, will close the issue(s) when PR gets merged): Fixes #
https://github.com/kubernetes-sigs/cloud-provider-azure/issues/5231
**Additional context**
Add any other context for the reviewers

### CODE REVIEW

File: images/capi/ansible/roles/providers/tasks/azure.yml [hunk 1]
```diff
@@ -79,3 +79,21 @@
     group: root
     mode: "0644"
   when: ansible_os_family == "Debian"
+
+- name: Create the credential provider path
+  file:
+    path: /var/lib/kubelet/credential-provider
+    state: directory
+    mode: 0755
```
Comment (AverageMarcus) [inline images/capi/ansible/roles/providers/tasks/azure.yml:87]:
```suggestion
    mode: "0755"
```

File: images/capi/ansible/roles/providers/tasks/azure.yml [hunk 2]
```diff
@@ -79,3 +79,21 @@
...
+
+- name: Download OOT credential provider
+  get_url:
+    url: https://github.com/kubernetes-sigs/cloud-provider-azure/releases/latest/download/azure-acr-credential-provider-linux-amd64
+    dest: /var/lib/kubelet/credential-provider/acr-credential-provider
+    mode: 0755
```
Comment (AverageMarcus) [inline images/capi/ansible/roles/providers/tasks/azure.yml:93]:
```suggestion
    mode: "0755"
```

File: images/capi/ansible/roles/providers/tasks/azure.yml [hunk 3]
```diff
@@ -79,3 +79,21 @@
...
+
+- name: Download OOT credential provider config file
+  get_url:
+    url: https://raw.githubusercontent.com/kubernetes-sigs/cloud-provider-azure/master/examples/out-of-tree/credential-provider-config.yaml
+    dest: /var/lib/kubelet/credential-provider-config.yaml
+    mode: 0644
```
Comment (AverageMarcus) [inline images/capi/ansible/roles/providers/tasks/azure.yml:99]:
```suggestion
    mode: "0644"
```

File: images/capi/ansible/roles/providers/tasks/azure.yml [hunk 4]
```diff
@@ -79,3 +79,21 @@
...
+
+- name: Download OOT credential provider
+  get_url:
+    url: https://github.com/kubernetes-sigs/cloud-provider-azure/releases/latest/download/azure-acr-credential-provider-linux-amd64
+    dest: /var/lib/kubelet/credential-provider/acr-credential-provider
+    mode: 0755
+
+- name: Download OOT credential provider config file
+  get_url:
+    url: https://raw.githubusercontent.com/kubernetes-sigs/cloud-provider-azure/master/examples/out-of-tree/credential-provider-config.yaml
```
Comment (AverageMarcus) [inline images/capi/ansible/roles/providers/tasks/azure.yml:97]: Is there any concern with always pointing at the latest `master` version of this? Do we maybe need to have the version configurable? (Same with the above download I guess)

File: images/capi/ansible/windows/roles/providers/tasks/azure.yml [hunk 5]
```diff
@@ -60,3 +60,22 @@
...
+    path: C:\var\lib\kubelet\credential-provider
+    state: directory
+
+- name: Download OOT credential provider
+  win_get_url:
+    url: https://github.com/kubernetes-sigs/cloud-provider-azure/releases/latest/download/azure-acr-credential-provider-windows-amd64.exe
+    dest: C:\var\lib\kubelet\credential-provider\acr-credential-provider.exe
+
+- name: Copy OOT credential provider binary
+  when: kubernetes_semver is version('v1.30.0', '<')
```
Comment (AverageMarcus) [inline images/capi/ansible/windows/roles/providers/tasks/azure.yml:75]: What changes in `v1.30` that means this is no longer needed? (And why only on Windows?)

**Review by AverageMarcus — COMMENTED** [review]
I'll leave the approval for @mboersma but I've added a few small comments. Mainly questions for my benefit.

Also... what does OOT stand for? My searching came up with nothing that seemed to fit. 😕 

/assign mboersma

File: images/capi/ansible/roles/providers/tasks/azure.yml (hunk 1 above)
Comment (lzhecheng) [reply images/capi/ansible/roles/providers/tasks/azure.yml]: Updated.

File: images/capi/ansible/roles/providers/tasks/azure.yml (hunk 2 above)
Comment (lzhecheng) [reply images/capi/ansible/roles/providers/tasks/azure.yml]: Updated.

File: images/capi/ansible/roles/providers/tasks/azure.yml (hunk 3 above)
Comment (lzhecheng) [reply images/capi/ansible/roles/providers/tasks/azure.yml]: Updated.

File: images/capi/ansible/roles/providers/tasks/azure.yml (hunk 4 above)
Comment (lzhecheng) [reply images/capi/ansible/roles/providers/tasks/azure.yml:97]: The fact is that this credential provider code and the config file almost don't change so far... I put it latest here but open to suggestions.

File: images/capi/ansible/windows/roles/providers/tasks/azure.yml (hunk 5 above)
Comment (lzhecheng) [reply images/capi/ansible/windows/roles/providers/tasks/azure.yml:77]: There's a bug for k8s v1.30- and the issue has been resolved: https://github.com/kubernetes-sigs/cloud-provider-azure/issues/4533
I added a comment including this link.

**Comment by lzhecheng** [comment]
@AverageMarcus OOT means [out of tree](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/2133-out-of-tree-credential-provider/README.md).

File: images/capi/ansible/roles/providers/tasks/azure.yml (hunk 4 above)
Comment (AverageMarcus) [reply images/capi/ansible/roles/providers/tasks/azure.yml:97]: As long as it always supports N-3 versions of Kubernetes I suspect it should be ok.

**Comment by AverageMarcus** [comment]
> @AverageMarcus OOT means [out of tree](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/2133-out-of-tree-credential-provider/README.md?rgh-link-date=2024-02-05T11%3A16%3A47Z).

Oooooohhhhhhhh! 🤦  I was thinking it was some Azure specific service.

**Review by mboersma — APPROVED** [review]
/lgtm

/assign @jsturtevant

**Pull request merged**